docs: update README (#1286)
xorq-labs/xorq: README.md

![PyPI - Version](https://img.shields.io/pypi/v/xorq)
![CI Status](https://img.shields.io/github/actions/workflow/status/xorq-labs/xorq/ci-test.yml)

**The periodic table for ML computation.**

Everything is an expression. Addressable. Composable. Portable.

Write high-level expression. Execute as SQL on DuckDB, Snowflake, BigQuery, or
any engine. Every computation addressable, versioned, and reusable.

[Documentation](https://docs.xorq.dev) • [Discord](https://discord.gg/8Kma9DhcJG) • [Website](https://www.xorq.dev)

</div>

---

## What is Xorq?

> ML infrastructure is fragmented—features in one system, models in another,

Reviewer: ```suggestion
Machine learning (ML) infrastructure is fragmented—features in one system, models in another,
```

lineage reconstructed through archaeology.

What if features, models, and pipelines aren't different things?

A feature is a computation. A model is a computation. A pipeline is
computations composed. The vendor categories aren't computational
truths—they're commercial territories. Strip away the product boundaries and
everything reduces to the same primitive: the expression.

Xorq is the composability layer for compute expressed as relational plans.

## Installation

```bash
pip install xorq[examples]
xorq init -t penguins
```

> [Full Tutorial →](https://docs.xorq.dev/tutorials/getting_started/quickstart)

Reviewer: Write this as part of a full sentence, link text should exactly match the target title.

## Quick Start

```python
import xorq.api as xo
from sklearn.ensemble import RandomForestClassifier

data = xo.read_parquet('s3://bucket/penguins.parquet')
train, test = xo.test_train_splits(data, test_size=0.2)

model = xo.Pipeline.from_instance(RandomForestClassifier())
fitted = model.fit(train, features=['bill_length_mm', 'bill_depth_mm'],
                   target='species')

predictions = fitted.predict(test).cache(storage=ParquetStorage())  # deferred
predictions.execute()  # do work
```

**CLI:**

```bash
xorq build expr.py -e predictions
xorq run builds/
```

> ## How It Works

Reviewer: ```suggestion
## How it works
```

Xorq captures your ML computation as an **input-addressed manifest**—a
declarative representation where each node is identified by the hash of its
computation specification, not its results.

```yaml
# Manifest snippet: fit → predict lineage
predicted:
  op: ExprScalarUDF            # Model inference
  kwargs:
    bill_length_mm: ...        # Feature inputs
    bill_depth_mm: ...
  meta:
    __config__:
      computed_kwargs_expr:    # Training lineage preserved
        op: AggUDF             # Model training
        kwargs:
          species: ...         # Original training target
```

### What This Enables

| Capability | How |
|------------|-----|
| **Version by intent** | Same computation = same hash, regardless of input data |
| **Precise caching** | Cache based on what you're computing, not when |
| **Structural lineage** | Provenance is the graph itself, not reconstructed logs |
| **Portable execution** | Manifest compiles to optimized SQL for any engine |

### Input-Addressing

Every computation gets a unique hash based on its logic:

- Same feature engineering on different days → **same hash** (reusable)
- Different feature logic → **different hash** (new version)

> If anyone on your team has run this exact computation before, Xorq reuses it

Reviewer: ```suggestion
If anyone on your team has run this exact computation before, then Xorq reuses it
```

automatically. The hash is the truth.

## The Catalog

Your team's shared ledger of ML compute—versioned, discoverable, composable:

> ```bash

Author: added context...

# Register a build with an alias
❯ xorq catalog add builds/7061dd65ff3c --alias fraud-model

# Discover what exists
❯ xorq catalog ls
Aliases:
fraud-model                  7061dd65ff3c     r2
customer-features            dbf90860-88b3    r1
recommendation-pipeline      52f987594254     r1

# Trace lineage
❯ xorq lineage fraud-model

# Serve for inference
xorq serve-unbound  fraud-model --port 8001 405154f690d20f4adbcc375252628b75
```

The catalog isn't a database. It's an addressing system—discoverable by humans,
navigable by agents.

> ## The Architecture

Author: Will do with the updated graphic

</div>

> **Xorq is a batch transformation framework built on Ibis,
> DataFusion and Arrow.**
> It ships a multi-engine manifest that you can run in SQL across DuckDB,
> Snowflake, DataFusion, and more.

---

## What Xorq gives you

| Feature | Description |
|---|---|
|**Multi-engine manifest** | A single, typed plan (YAML manifest) that executes as SQL on DuckDB, Snowflake, and embedded DataFusion. |

> |**Deterministic builds & caching** | Manifests are uniquely named with a deterministic hash of the expression. |

Reviewer: ```suggestion
|**Deterministic builds and caching** | Manifests are uniquely named with a deterministic hash of the expression. |
```

> |**Lineage & schemas** | Schema checks with end-to-end, column-level lineage. |

Reviewer: ```suggestion
|**Lineage and schemas** | Schema checks with end-to-end, column-level lineage. |
```

|**Compute catalog** | Versioned registry to run, cache, diff, and manifests. |
|**Portable UDxFs** | Arbitrary Python logic with schema-in/out contracts, portable via Arrow Flight. |
|**`scikit-learn` integration** | Fit/predict pipelines serialize to a manifest for portable batch scoring with training lineage. |
|**Templates with `uv`** | `xorq init` ships a templates in **replicaple environments** |

> [!NOTE]

> > **Not an orchestrator.** Use Xorq from Airflow, Dagster, GitHub Actions, etc.

Reviewer: ```suggestion
> **Not an orchestrator.** Use Xorq from Airflow, Dagster, GitHub Actions, and more.
```

> **Xorq is a multi‑engine batch transformation framework built on Ibis,
> DataFusion and Arrow.**
> It ships a multi-engine manifest that you can run in SQL across DuckDB,
> Snowflake, DataFusion, and more.

---

## What Xorq gives you

| Feature | Description |
|---|---|
|**Multi-engine manifest** | A single, typed plan (YAML manifest) that executes as SQL on DuckDB, Snowflake, and embedded DataFusion. |

> |**Deterministic builds & caching** | One hash for everything—computed from **expression inputs**; for YAML-only builds, we hash the **expression**. The hash names `builds/<hash>/` and keys the cache. |

Reviewer 2: `Manifests are uniquely named with a deterministic hash of the expression`

> |**Lineage & schemas** | Compile-time schema checks with end-to-end, column-level lineage. |

Reviewer 2: schema checks occur during the building of the expr in memory

Tutorial](https://docs.xorq.dev/tutorials/getting_started/quickstart) for a
full walk-through using the Penguins dataset.

### Project Templates

We ship minimal, opinionated starter templates so you can go from
zero-to-manifest fast.

- **Penguins:** Feature engineering + fit/predict LogisticRegression on the
Penguins dataset.
- **Digits:** Fit/predict on the Digits dataset with a full pipeline (PCA +
classifier).

Each template includes:

```bash
uv.lock — pinned dependencies for replicable envs
requirements.txt — bare minimal requirement
pyproject.toml — project metadata
expr.py  — the expr entrypoint
```

#### Requirements for environment replicability for a Project:
- TBD

> ## Multi-engine manifest for Machine Learning pipelines

Reviewer: ```suggestion
## Multi-engine manifest for machine learning pipelines
```

          body_mass_g: ...
          species: ...         # target
```

> We serialize the expression as a YAML manifest that captures the graph and all

Reviewer: Who is "we"?

          body_mass_g: ...
          species: ...         # target
```

We serialize the expression as a YAML manifest that captures the graph and all

> nodes (including UDFs as pickled entries); builds are content-addressed by the

Author: are build hashes also input-addressed?

A generic catalog that can be used to build new workloads:

- ML/data pipeline development (deterministic builds, caching, replicable envs)
- Lineage‑preserving, multi-engine feature stores (offline, reproducible)
- Composable data products (ship datasets as compute artifacts)
- Governed sharing of compute (catalog entries as the contract between teams)

Also great for:

> - Generating SQL from high-level DSLs (e.g. Semantic Layers)

Reviewer: Avoid using e.g. and i.e.
